extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js

```diff
@@ -1,30 +1,45 @@
-import {Layout, applyFillContent} from '#core/dom/layout';
+import {Layout} from '#core/dom/layout';
 
-const TAG = 'amp-story-shopping-tag';
+import {Services} from '#service';
+
+import {StateProperty} from '../../amp-story/1.0/amp-story-store-service';
 
 export class AmpStoryShoppingTag extends AMP.BaseElement {
   /** @param {!AmpElement} element */
   constructor(element) {
     super(element);
 
-    /** @private {string} */
-    this.myText_ = TAG;
-
-    /** @private {?Element} */
-    this.container_ = null;
+    /** @private @const {!./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = null;
   }
 
   /** @override */
   buildCallback() {
     super.buildCallback();
-    this.container_ = this.element.ownerDocument.createElement('div');
-    this.container_.textContent = this.myText_;
-    this.element.appendChild(this.container_);
-    applyFillContent(this.container_, /* replacedContent */ true);
+    return Promise.all([
+      Services.storyStoreServiceForOrNull(this.win).then((storeService) => {
+        this.storeService_ = storeService;
+        this.storeService_.subscribe(
+          StateProperty.SHOPPING_STATE,
+          (shoppingState) => this.updateShoppingTag_(shoppingState)
+        );
+      }),
+    ]);
   }
 
   /** @override */
   isLayoutSupported(layout) {
     return layout === Layout.CONTAINER;
   }
+
+  /**
+   * @param {Object} shoppingState
+   * @private
+   */
+  updateShoppingTag_(shoppingState) {
+    const tagData = shoppingState[this.element.getAttribute('data-tag-id')];
+    if (tagData != null) {
+      this.element.textContent = tagData['product-title'];
```

Comment:
Any DOM mutation needs to be in a `this.measureElement()` callback so we get good browser frame management, and avoid layout thrashing

Reply:
Chatted offline with @gmajoulet and confirmed that he actually meant to say that DOM mutations need to be done in a `this.mutateElement()`.

Reply:
Okay, thanks for clarifying Corey.